enumerable.rb

```diff
@@ -0,0 +1,125 @@
+# rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/ModuleLength,
+
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    n = 0
+    while n < to_a.length
+      yield(self[n]) if is_a?(Array)
+      yield(keys[n], [keys[n]]) if is_a?(Hash)
+      yield(to_a[n]) if is_a?(Range)
+      n += 1
+    end
+    self
+  end
+
+  def my_each_with_index(&block)
+    return to_enum(:my_each_with_index) unless block_given?
+
+    n = 0
+    while n < to_a.length
+      block.call(self[n], n) if is_a?(Array)
+      block.call(keys[n], [keys[n]]) if is_a?(Hash)
+      block.call(to_a[n]) if is_a?(Range)
+      n += 1
+    end
+    self
+  end
```

Comment:
**`my_each_with_index` when iterated over range prints index values as nil.**

```
hash = Hash.new

(3..7).each_with_index { |item, index|
  hash[item] = index
}
print hash #=> {3=>0, 4=>1, 5=>2, 6=>3, 7=>4}

(3..7).my_each_with_index { |item, index|
  hash[item] = index
}

print hash #=> {3=>nil, 4=>nil, 5=>nil, 6=>nil, 7=>nil}
```
**`my_each_with_index`  does not give the correct output when iterated over hash values.**

```
hash = Hash.new

{foo: 0, bar: 1, baz: 2}.each_with_index { |item, index|
  hash[item] = index
}

print hash #=> {[:foo, 0]=>0, [:bar, 1]=>1, [:baz, 2]=>2}

{foo: 0, bar: 1, baz: 2}.my_each_with_index { |item, index|
  hash[item] = index
}

print hash #=> {:foo=>[:foo], :bar=>[:bar], :baz=>[:baz]}
```